### PR TITLE
Api server can call yagna daemon

### PR DESCRIPTION
**Pull request by pnowosie**
Delivers [[APPS-78](https://golemproject.atlassian.net/browse/APPS-78)], subtask of [[APPS-33](https://golemproject.atlassian.net/browse/APPS-33)]

Provides a docker-compose file with simplifies image creation and containers startup. Yagna service data dir is linked to `server/yagna_data` directory.

When api server's container is started, yagna service might not be ready yet (requestor account needs to be unlocked, funded and app-key generated). Api will serve incoming requests but also is running background task that is aim to first [obtain app-key](https://github.com/golemfactory/yagna-service-poc/blob/3c873c29a12c413e73b50690ab435305ace86574/server/api/src/delayed_init.py#L9) and using the key, fetch the [requestor accounts](https://github.com/golemfactory/yagna-service-poc/blob/3c873c29a12c413e73b50690ab435305ace86574/server/api/src/delayed_init.py#L30) from yagna rest service.

I'll provide some sdk code in the further task...

### Port bindings
| Service | URL |
| :--------- | ------ |
| API server | `localhost:5000` |
| YAGNA REST | `localhost:6000` |
| YAGNA GSB | `localhost:6010` |

## DIY
```
cd server
make up

# containers shutdown
make down

# clean docker artifacts
make clean
```

After starting the containers please check when API is already up & ready by calling (undocumented) 
`GET localhost:5000/api`
```
{
  "git.branch": "pnowosie/apiserver-calls-yagna-APPS-78", 
  "git.commit": "3c873c2", 
  "version": "0.0.1", 
  "yagna.available": true,     <- key was obtained
  "yagna.ready": true          <- got result from rest
}
```

:ok_hand: you can restart (down & up) which is faster as it's reuses the account (but some hardcoded sleeps still applies :open_hands: )

## Embracing async API

**Out-dated, read update**

I moved from old, good Flask to Quart which supports Python's asyncio. The transition was smooth, just [replace `flask|quart`](https://github.com/golemfactory/yagna-service-poc/pull/5/commits/ac21514df8163b8965ccf65654433bd20399d130). But then it complicates things (which might be due my inexperience with python / asyncio). 
(...)

### Update

I've read about Quart a little bit more. As the result of the knowledge gained there is a proper implementation of waiting on Yagna. I've implemented [background tasks](https://github.com/golemfactory/yagna-service-poc/blob/3c873c29a12c413e73b50690ab435305ace86574/server/api/src/app.py#L41) that:
1. Waits for `app_key` file generated by Yagna container init script in the common volumen
2. Parses the app_key to obtain Yagna's rest token (app-key)
3. Calls Yagna's requestor accounts to compare address fetched from the app key

Here is how the new start-up looks like
```
╰─ docker logs -f server_api_1                                                                                   ─╯
 * Serving Quart app 'app'
 * Environment: development
 * Debug mode: True
 * Running on http://0.0.0.0:5000 (CTRL + C to quit)
Waiting for Yagna service ready... (allowing 60 secs)
[2021-04-23 21:07:43,042] Running on http://0.0.0.0:5000 (CTRL + C to quit)
Checking for app_Key file...
[2021-04-23 21:07:48,047] 172.29.0.1:58944 GET /api 1.1 200 165 5486
Checking for app_Key file...
Checking for app_Key file...
[2021-04-23 21:07:53,562] 172.29.0.1:58954 GET /api 1.1 200 165 4062
Checking for app_Key file...
Checking for app_Key file...
Checking for app_Key file...
Checking for app_Key file...
Yagna started! yagna_account=0x17cf1add0d85fc1d76717bf126f4184d87eba0d2, yagna_app_key=5110abf0ea334219aa5e67dc5e0d41f0
Trying to get requestors response=[{'platform': 'zksync-rinkeby-tglm', 'address': '0x17cf1add0d85fc1d76717bf126f4184d87eba0d2', 'driver': 'zksync', 'network': 'rinkeby', 'token': 'tGLM', 'send': True, 'receive': True}]
[2021-04-23 21:08:07,525] 172.29.0.1:58968 GET /api 1.1 200 163 4600

```

### CODE REVIEW

File: server/api/src/app.py [hunk 1]
```diff
@@ -1,27 +1,43 @@
...
+    async def classify():
+        form = await request.form
+        text = form.get("text_file") or form.get("text")
         print(text)
 
         return classifier.classify_text(text)
 
+    @app.before_serving
+    async def init_wait_yagna():
+        if not hasattr(app, "yagna_available"):
```
Comment (pnowosie) [inline server/api/src/app.py:35]: Not need to check here - it's just called at the beginning, before any request  gets handled

File: server/docker/Makefile [hunk 2]
```diff
@@ -1,23 +0,0 @@
-# Makefile is your friend!
```
Comment (pnowosie) [inline server/docker/Makefile:1]: Note: `docker` folder has been renamed to `yagna`

File: server/api/src/delayed_init.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
+import asyncio
+import aiofiles
+import aiohttp
+import os
+
+from utils import parse_yagna_key
+from time import time
+
+async def is_yagna_available(app):
```
Comment (pnowosie) [inline server/api/src/delayed_init.py:9]: Maybe I shouldn't pass the `app` but smth smaller 🤔

**Review by albatrosica — APPROVED** [review]
No server-related problems detected with this config 👍

File: server/api/src/delayed_init.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    global YAGNA_ACCOUNT, YAGNA_APP_KEY
```
Comment (Wiezzel) [inline server/api/src/delayed_init.py:10]: Why not make these into `app` attributes like `yagna_available`? Maybe `yagna_available` won't even be needed anymore if we can check these two?

File: server/api/src/utils.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
+import json
+
+
+def parse_yagna_key(response_json: str) -> (str, str):
```
Comment (Wiezzel) [inline server/api/src/utils.py:4]: Unit tests for this function are highly welcome.

File: server/yagna/init.sh [hunk 6]
```diff
@@ -13,6 +13,10 @@ sleep 5
 /usr/bin/yagna payment fund --driver zksync
 /usr/bin/yagna payment init --sender --driver zksync
 
+# having key generation a last step allows api service to wait until yagna is ready
+/usr/bin/yagna app-key create requestor
```
Comment (Wiezzel) [inline server/yagna/init.sh:17]: This command outputs the generated key. You could save this and skip the whole JSON parsing.

File: server/api/src/utils.py (hunk 5 above)
Comment (Wiezzel) [reply server/api/src/utils.py:4]: But probably it's not needed at all – see comment on `server/yagna/init.sh`

File: server/yagna/init.sh (hunk 6 above)
Comment (pnowosie) [reply server/yagna/init.sh:17]: Yes, actually this was [the 1st approach](https://github.com/golemfactory/yagna-service-poc/pull/5/commits/b454a6afeb83ea0d3a5d1288cfed7f5b20c8abdb#diff-654ade571040d742e696c7195dae43b8d7de4f328b4d31f9789db308da6240ac) however the file's content will be erased on subsequent run (starting yagna again on previously generated files)

File: server/api/src/utils.py (hunk 5 above)
Comment (pnowosie) [reply server/api/src/utils.py:4]: Tests added in https://github.com/golemfactory/yagna-service-poc/commit/e88222eb8a8c9bd7b44a88ca1eccf277c20ecc19

File: server/api/src/delayed_init.py (hunk 4 above)
Comment (pnowosie) [reply server/api/src/delayed_init.py:10]: I not yet sure whether I would need them (or APPKEY) have in env vars so I will postpone the change to the next, further iteration

**Review by Wiezzel — APPROVED** [review]

**Comment by johny-b** [comment]
I didn't think this through, but just an idea.
There is a lot of things going on with `is-yagna-ready`. I could think of two different solutions:
A) something like this in `CMD` in api/Dockerfile:
```
CMD wait_for_yagna_ready.sh && QUART_ENV=development QUART_APP="app:create_app()" quart run --host=0.0.0.0
```
(this will not work, but either `bash -c` or separate `init.sh` script will do)

(where `wait_for_yagna_ready.sh` just sleeps untill yagna answers).
I was using something like `wait-for-postgres` in similar scenario and this was something like:
```
while ! curl http://$POSTGRES_PORT_5432_TCP_ADDR:$POSTGRES_PORT_5432_TCP_PORT/ 2>&1 | grep '52'
do
  sleep 1
done
```
and it worked.

B) you could add in `api` service definition
```
    depends_on:
      yagna:
        condition: service_healthy
 ```
 and in `yagna` define appropriate healthcheck (note that healthchecks are useful either way), something like:
 
 ```
     healthcheck:
        test: ["yagna", "status", "something"]  # command that returns 0 if yagna is available and something else if not
        interval: 30s
        timeout: 10s
        retries: 5
```
    
    
I like both solutions more than the current, because they handle "waiting for yagna" thing to a separate place, and also there is no `asyncing` in them : )

(disclaimer: I didn't analyze the topic thoroughly)

**Review by johny-b — COMMENTED** [review]
Looks good, just check the comment above : )

**Comment by albatrosica** [comment]
Please add readme file with instructions how to make server running.

**Comment by pnowosie** [comment]
@johny-b Thank you! Good insights.
Both above propositions lacks ability to return the status to the Frontend, which might be important as it puts not constraints how we host the solution.

(BTW I thied [`A` here](https://github.com/golemfactory/yagna-service-poc/pull/5/commits/ce54b552024a5b9cbf81110f5068b7531462de6a), disclaimer: code smell 🐽 )

However `B` is something we might consider, if we stick to the containerized deployment (on single machine)
`Frontend (depends on) Backend (depends on) Yagna`
and the frontend container starts last.

**Comment by pnowosie** [comment]
> Please add readme file with instructions how to make server running.

I'm thinking about one `docker-compose.yml` that starts everything. But for now - because it's changing fast - I'm about to postpone the proper instructions.

**Pull request merged**